src/validations/html/custom.js

```diff
@@ -0,0 +1,144 @@
+  let unopenedTagQueuesByName = Map.of();
+
+  parser.on('startTag', (name, attrs, selfClosing, location) => {
+    if (!(selfClosing || name in voidElements)) {
+      openTagStack = openTagStack.push({
+        name,
+        location: new Location(location),
+      });
+    }
+  });
+
+  parser.on('endTag', (name, locationObject) => {
+    const location = new Location(locationObject);
+
+    const openIndex = openTagStack.findLastIndex(
+      openTag => openTag.name === name);
+    const opened = openIndex >= 0;
+
+    // Assume any open tags after the matching open tag are mismatched. If the
+    // source is valid HTML, then no other open tags should exist. If no
+    // matching open tag exists, then all open tags are mismatched.
+    const mismatchedTags =
+      opened ? openTagStack.slice(openIndex + 1) : openTagStack;
+    mismatchedTagErrorsByLocation = mismatchedTagErrorsByLocation.set(
+      location, {name, mismatchedTags});
+
+    if (opened) {
+      openTagStack = openTagStack.slice(0, openIndex);
+    } else {
+      unopenedTagLocations = unopenedTagLocations.add(location);
+      unopenedTagQueuesByName =
+        unopenedTagQueuesByName.update(
+          name, List.of(), tags => tags.push({name, location}));
+
+      // If no matching open tag exists, then assume later close tags could
+      // match some open tag instead of emptying openTagStack.
+    }
+  });
+
+  return function* () {
+    // Suggest misplaced-close-tag if a spare unopened tag exists. If no spare
+    // exists:
+    //
+    // A mismatched close tag that has been opened means the case
+    //
+    // <div>
+    //   <span>
+    // </div>
+    //
+    // so suggest unclosed-tag-by-mismatched-close-tag.
+    //
+    // A mismatched close tag that has yet to be opened means the case
+    //
+    //   <span>
+    //   </span>
+    // </div>
+    //
+    // so suggest unexpected-close-tag.
+    for (const [location, {name, mismatchedTags}]
+         of mismatchedTagErrorsByLocation) {
+      const opened = !unopenedTagLocations.has(location);
+      for (const mismatchedTag of mismatchedTags) {
+        const spareUnopenedTag =
+          unopenedTagQueuesByName.get(mismatchedTag.name, List.of()).first();
+        if (spareUnopenedTag) {
+          unopenedTagQueuesByName = unopenedTagQueuesByName.update(
+            spareUnopenedTag.name, tags => tags.shift());
+
+          yield {
+            row: location.get('line') - 1,
+            column: location.get('col') - 1,
+            reason: 'misplaced-close-tag',
+            payload: {
+              open: mismatchedTag.name,
+              close: name,
+              misplacement: spareUnopenedTag.location.line,
+            },
+            suppresses: ['unexpected-close-tag'],
+          };
+        } else if (opened) {
+          yield {
+            row: location.get('line') - 1,
+            column: location.get('col') - 1,
+            reason: 'unclosed-tag-by-mismatched-close-tag',
```

Comment:
Is there a reason not to just report this as a normal `unclosed-tag`? Is it really useful to include information about where we found a mismatched closing tag? Would it make more sense to still locate the error at the location of the opening tag? (These are all sincere/non-rhetorical questions!)